### PR TITLE
Updated to the most recent Alonzo specification

### PR DESCRIPTION
**Pull request by TimSheard**
Completed Figures 1, 5, 6, 7, 8 and 12. Ormolise.

This PR implements all the functions in the most up todate specification in the above Figures.

### CODE REVIEW

File: alonzo/impl/src/Cardano/Ledger/Alonzo/PParams.hs [hunk 1]
```diff
@@ -24,30 +27,30 @@ module Cardano.Ledger.Alonzo.PParams
...
 
-import Cardano.Binary
-  ( FromCBOR (..),
-    ToCBOR (..),
-    decodeWord,
-    encodeListLen,
-    encodeMapLen,
-    encodeWord,
-  )
+import Cardano.Binary (FromCBOR (..), ToCBOR (..), decodeWord, encodeListLen, encodeMapLen, encodeWord, serialize')
```
Comment (JaredCorduan) [inline alonzo/impl/src/Cardano/Ledger/Alonzo/PParams.hs]: maybe it's just me, but when the imports get past three or so, I prefer them vertical. it makes it easier to see what is changing in diffs.

File: alonzo/impl/src/Cardano/Ledger/Alonzo/PParams.hs [hunk 2]
```diff
@@ -413,3 +420,39 @@ newtype PPHash crypto
...
+    f (HashItem x) = serialize' x
+
+-- Hash with 2 Proxys. This function is only used internally
+-- The type families need alot of help keepng things straight,
+-- the Proxys (and the type signature) provide that help.
+
+hash2P :: forall t e. Era e => Proxy t -> Proxy e -> [HashItem] -> Hash.Hash (CC.HASH (Crypto e)) t
+hash2P _ _ x = Hash.castHash (Hash.hashWith serializeHashTuple x)
+
+class Era e => Hashable e t where
```
Comment (JaredCorduan) [inline alonzo/impl/src/Cardano/Ledger/Alonzo/PParams.hs]: we are using the `Hashable` typeclass from `Data.Hashable` (hashable package) elsewhere in our codebase, so we will need a new name.

File: alonzo/impl/src/Cardano/Ledger/Alonzo/Tx.hs [hunk 3]
```diff
@@ -17,37 +22,80 @@ module Cardano.Ledger.Alonzo.Tx
     Tx (Tx, body, wits, isValidating, auxiliaryData),
     ScriptPurpose (..),
     Indexable (..),
+    --  Figure 5
```
Comment (JaredCorduan) [inline alonzo/impl/src/Cardano/Ledger/Alonzo/Tx.hs:50]: this is awesome lining up these exports with the figures from the spec!

File: alonzo/impl/src/Cardano/Ledger/Alonzo/Tx.hs [hunk 4]
```diff
@@ -204,23 +252,92 @@ deriving via
...
+
+-- | The keys of all the inputs of the TxBody (both the inputs for fees, and the normal inputs).
+txins :: AlonzoBody era => TxBody era -> Set (TxIn (Crypto era))
+txins (TxBody {inputs = is, inputs_fee = fs}) = Set.union is fs
+
+txscriptfee :: Prices -> ExUnits -> Coin
+txscriptfee (Prices pr_mem pr_steps) (ExUnits mem steps) =
+  (mem <×> pr_mem) <+> (steps <×> pr_steps)
+
+-- TODO Check that txSize is the right functio-- Length of the serialised bytes
```
Comment (JaredCorduan) [inline alonzo/impl/src/Cardano/Ledger/Alonzo/Tx.hs]: yep, what you defined below is correct. note that in shelley we [didn't camel case](https://github.com/input-output-hk/cardano-ledger-specs/blob/d104b97739bf37b695251b3ece5430f867fe0411/shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/LedgerState.hs#L772) it (and we should be consistent).

File: alonzo/impl/src/Cardano/Ledger/Alonzo/Tx.hs [hunk 5]
```diff
@@ -204,23 +252,92 @@ deriving via
...
+minfee pp tx =
+  ((txSize tx) <×> (a pp))
+    <+> (b pp)
+    <+> (txscriptfee (_prices pp) (exunits (txbody tx)))
+  where
+    a protparam = Coin (fromIntegral (_minfeeA protparam))
+    b protparam = Coin (fromIntegral (_minfeeB protparam))
+
+-- IT IS NOT AT ALL CLEAR what "validatorHash a" means in
+-- the specification. Maybe something like this?
```
Comment (JaredCorduan) [inline alonzo/impl/src/Cardano/Ledger/Alonzo/Tx.hs]: yes, this is correct. let's remove this comment before merging. we could add some prose to the spec if you think that that would help.

File: alonzo/impl/src/Cardano/Ledger/Alonzo/Tx.hs [hunk 6]
```diff
@@ -261,7 +378,193 @@ indexedRdmrs ::
...
+      Nothing -> []
+      Just txout ->
+        case getField @"datahash" txout of
+          Nothing -> []
+          Just hash ->
+            case Map.lookup hash (witsData (wits tx)) of
+              Nothing -> []
+              Just d -> [d]
+
+-- IS IT REALLY THE CASE ONLY SPENDING PURPOSE HAS DATA? TODO
```
Comment (JaredCorduan) [inline alonzo/impl/src/Cardano/Ledger/Alonzo/Tx.hs]: maybe @polinavino  or @WhatisRT can confirm.

File: alonzo/impl/src/Cardano/Ledger/Alonzo/Tx.hs [hunk 7]
```diff
@@ -261,7 +378,193 @@ indexedRdmrs ::
...
+  [(ScriptPurpose crypto, ScriptHash crypto)] ->
+  DCert crypto ->
+  [(ScriptPurpose crypto, ScriptHash crypto)]
+add_only_cwitness !ans !(DCertDeleg (c@(DeRegKey (ScriptHashObj hk)))) =
+  (Certifying $ DCertDeleg c, hk) : ans
+add_only_cwitness !ans !(DCertDeleg (c@(Delegate (Delegation (ScriptHashObj hk) _dpool)))) =
+  (Certifying $ DCertDeleg c, hk) : ans
+add_only_cwitness !ans _ = ans
+
+-- TODO   Why is 'h' in the existential? h should be underscore or _h
```
Comment (JaredCorduan) [inline alonzo/impl/src/Cardano/Ledger/Alonzo/Tx.hs]: we can remove this comment now.

File: alonzo/impl/src/Cardano/Ledger/Alonzo/TxBody.hs [hunk 8]
```diff
@@ -98,6 +94,7 @@ newtype IsFee = IsFee Bool
       FromCBOR
     )
 
+{-
```
Comment (JaredCorduan) [inline alonzo/impl/src/Cardano/Ledger/Alonzo/TxBody.hs]: since `TxIn` will be the same in Shelley and Alonzo, can't we delete this commented code?

**Review by JaredCorduan — APPROVED** [review]
This is a bunch of the alonzo figures, all lining up so nicely with the spec, it's looking really great! I made a few comments about mostly trivial things.

File: alonzo/impl/src/Cardano/Ledger/Alonzo/Tx.hs (hunk 6 above)
Comment (WhatisRT) [reply alonzo/impl/src/Cardano/Ledger/Alonzo/Tx.hs]: yes, all the other purposes are essentially state-less, so they don't have data

File: alonzo/impl/src/Cardano/Ledger/Alonzo/PParams.hs [hunk 9]
```diff
@@ -75,6 +78,7 @@ import Shelley.Spec.Ledger.Serialization
   )
 import Shelley.Spec.Ledger.Slot (EpochNo (..))
 
+-- ================================================================
```
Comment (nc6) [inline alonzo/impl/src/Cardano/Ledger/Alonzo/PParams.hs:93]: Very minor, but I find these things kind of annoying, and I know for example that they end up being removed in other PRs. Do you find them that helpful?

File: alonzo/impl/src/Cardano/Ledger/Alonzo/PParams.hs [hunk 10]
```diff
@@ -401,6 +405,9 @@ updatePParams pp ppup =
     fromMaybe' :: a -> StrictMaybe a -> a
     fromMaybe' x = fromMaybe x . strictMaybeToMaybe
 
+-- ===================================================
+-- Figure 1: "Definitions Used in Protocol Parameters"
```
Comment (nc6) [inline alonzo/impl/src/Cardano/Ledger/Alonzo/PParams.hs:431]: I think if you're going to link to the doc, we need to make the link more explicit

File: alonzo/impl/src/Cardano/Ledger/Alonzo/PParams.hs [hunk 11]
```diff
@@ -413,3 +420,39 @@ newtype PPHash crypto
 deriving newtype instance CC.Crypto crypto => FromCBOR (PPHash crypto)
 
 deriving newtype instance CC.Crypto crypto => ToCBOR (PPHash crypto)
+
+hashLanguagePP :: forall era. Era era => PParams era -> Language -> PPHash (Crypto era)
+hashLanguagePP pp lang = PPHash (hash2P (Proxy @EraIndependentPP) (Proxy @era) [HashItem restrictedmap])
```
Comment (nc6) [inline alonzo/impl/src/Cardano/Ledger/Alonzo/PParams.hs]: Very long line, can we try to keep lines to 80 chars?

File: alonzo/impl/src/Cardano/Ledger/Alonzo/PParams.hs [hunk 12]
```diff
@@ -413,3 +420,39 @@ newtype PPHash crypto
...
+  where
+    restrictedmap :: Map.Map Language CostModel
+    restrictedmap = eval (singleton lang ◁ (_costmdls pp))
+
+-- ===============================================================
+-- Hashing Tuples (upto to width 4) in a uniform way
```
Comment (nc6) [inline alonzo/impl/src/Cardano/Ledger/Alonzo/PParams.hs]: I'm not convinced that this is something we want? We generally err on the side of _not_ doing things in such a uniform way (e.g. not using generic instances for serialisation) because we want control over how exactly we do so in areas which will affect hashes/validation of data stored on the chain.

We also don't use this functionality outside of this module, so it's not clear what the advantage of it is - it adds a new abstraction that somewhat duplicates existing abstractions but which isn't widely used.

File: alonzo/impl/src/Cardano/Ledger/Alonzo/Tx.hs [hunk 13]
```diff
@@ -17,37 +22,80 @@ module Cardano.Ledger.Alonzo.Tx
     Tx (Tx, body, wits, isValidating, auxiliaryData),
     ScriptPurpose (..),
     Indexable (..),
+    --  Figure 5
+    txinputs_fee,
```
Comment (nc6) [inline alonzo/impl/src/Cardano/Ledger/Alonzo/Tx.hs]: Why snake case? We typically reserve this for c interaction functions.

File: alonzo/impl/src/Cardano/Ledger/Alonzo/Tx.hs [hunk 14]
```diff
@@ -204,23 +252,92 @@ deriving via
     ) =>
     FromCBOR (Annotator (Tx era))
 
--- ===========================================
+-- ===============================================================
+-- From the specification, Figure 5 "Functions related to fees"
+-- ===============================================================
+
+isNonNativeScriptAddress :: forall era. ValidateScript era => Tx era -> Addr (Crypto era) -> Bool
```
Comment (nc6) [inline alonzo/impl/src/Cardano/Ledger/Alonzo/Tx.hs]: [style] line length

File: alonzo/impl/src/Cardano/Ledger/Alonzo/Tx.hs [hunk 15]
```diff
@@ -204,23 +252,92 @@ deriving via
...
+getValidatorHash (Addr _network (ScriptHashObj hash) _ref) = Just hash
+getValidatorHash _ = Nothing
+
+txbody :: Tx era -> TxBody era
+txbody (TxConstr (Memo (TxRaw {_body = b}) _)) = b
+
+txSize :: Tx era -> Integer
+txSize (TxConstr (Memo _ bytes)) = fromIntegral (SBS.length bytes)
+
+txinputs_fee :: AlonzoBody era => TxBody era -> Set (TxIn (Crypto era))
```
Comment (nc6) [inline alonzo/impl/src/Cardano/Ledger/Alonzo/Tx.hs]: This is just a field, we shouldn't need a new function for it

File: alonzo/impl/src/Cardano/Ledger/Alonzo/Tx.hs [hunk 16]
```diff
@@ -17,37 +22,80 @@ module Cardano.Ledger.Alonzo.Tx
...
+import Control.SetAlgebra (eval, (◁))
+import qualified Data.ByteString.Short as SBS (length)
 import Data.Coders
+import Data.List (foldl')
 import qualified Data.Map as Map
+import Data.Maybe (isJust, maybeToList)
 import Data.MemoBytes (Mem, MemoBytes (Memo), memoBytes)
 import Data.Sequence.Strict (StrictSeq)
 import qualified Data.Sequence.Strict as StrictSeq
-import qualified Data.Set as Set
```
Comment (nc6) [inline alonzo/impl/src/Cardano/Ledger/Alonzo/Tx.hs:42]: Data.Set recommends (https://hackage.haskell.org/package/containers-0.6.4.1/docs/Data-Set.html) qualified import, and this is consistent with the rest of the codebase. So please don't make it different here!

File: alonzo/impl/src/Cardano/Ledger/Alonzo/Tx.hs [hunk 17]
```diff
@@ -261,7 +378,193 @@ indexedRdmrs ::
...
+indexedRdmrs tx sp = Map.lookup policyid (txrdmrs . txwits $ tx)
   where
     policyid = rdptr (body tx) sp
+
+-- ===============================================================
+-- From the specification, Figure 7 "Script Validation, cont."
+-- ===============================================================
+
+-- | valContext collects info from the Tx and the UTxO and translates it into
+--   A Data, which the Plutus language know how to interpret.
```
Comment (nc6) [inline alonzo/impl/src/Cardano/Ledger/Alonzo/Tx.hs]:
```suggestion
--   a 'Data', which the Plutus language know how to interpret.
```

File: alonzo/impl/src/Cardano/Ledger/Alonzo/Tx.hs [hunk 18]
```diff
@@ -261,7 +378,193 @@ indexedRdmrs ::
...
+valContext :: UTxO era -> Tx era -> ScriptPurpose (Crypto era) -> [Data era]
+valContext _utxo _tx _sp = []
+
+--TODO FIX THIS, when defined will always return singleton list
```
Comment (nc6) [inline alonzo/impl/src/Cardano/Ledger/Alonzo/Tx.hs:492]: ?

File: alonzo/impl/src/Cardano/Ledger/Alonzo/Tx.hs [hunk 19]
```diff
@@ -261,7 +378,193 @@ indexedRdmrs ::
...
+  ]
+
+evalScripts :: (AlonzoScript.Script era, [Data era], ExUnits, CostModel) -> Bool
+evalScripts (AlonzoScript.NativeScript _timelock, _, _, _) = True
+evalScripts (AlonzoScript.PlutusScript, ds, units, cost) = b
+  where
+    (IsValidating b, _exunits) = runPLCScript cost AlonzoScript.PlutusScript ds units
+
+-- ===================================================================
+-- From Specification, Figure 12 "UTXOW helper functions"
```
Comment (nc6) [inline alonzo/impl/src/Cardano/Ledger/Alonzo/Tx.hs:562]: I feel like we might want to eventually move these elsewhere, to avoid `Tx` becoming a giant bag of all functions (a la `LedgerState.hs`), but we can do that subsequently

File: alonzo/impl/src/Cardano/Ledger/Alonzo/Tx.hs [hunk 20]
```diff
@@ -261,7 +378,193 @@ indexedRdmrs ::
...
+
+    !cert = foldl add_only_cwitness [] (certs txb)
+
+    !minted = map (\(pid@(PolicyID hash)) -> (Minting pid, hash)) (Map.keys m3)
+      where
+        m3 = getMapFromValue (mint txb)
+
+-- We only find certificate witnesses in Delegating and Deregistration DCerts
+-- that have ScriptHashObj credentials.
+add_only_cwitness ::
```
Comment (nc6) [inline alonzo/impl/src/Cardano/Ledger/Alonzo/Tx.hs]: snake_case?

File: alonzo/impl/src/Cardano/Ledger/Alonzo/Tx.hs [hunk 21]
```diff
@@ -261,7 +378,193 @@ indexedRdmrs ::
...
+-- languages:: TxWitness era -> Set Language  -- TODO
+
+txscripts ::
+  (Era era, ToCBOR (Core.Script era)) =>
+  TxWitness era ->
+  Map.Map (ScriptHash (Crypto era)) (Core.Script era)
+txscripts x = witsScript x
+
+txwits :: (Era era, ToCBOR (Core.AuxiliaryData era)) => Tx era -> TxWitness era
+txwits x = wits x
```
Comment (nc6) [inline alonzo/impl/src/Cardano/Ledger/Alonzo/Tx.hs:647]: Why this duplicate function?

File: alonzo/impl/src/Cardano/Ledger/Alonzo/Tx.hs [hunk 22]
```diff
@@ -261,7 +378,193 @@ indexedRdmrs ::
...
+-- hashSD :: TxWitness era -> Maybe (ScriptDataHash (Crypto era))
+-- is defined in Cardano.Ledger.Alonzo.TxWitness
+
+-- languages:: TxWitness era -> Set Language  -- TODO
+
+txscripts ::
+  (Era era, ToCBOR (Core.Script era)) =>
+  TxWitness era ->
+  Map.Map (ScriptHash (Crypto era)) (Core.Script era)
+txscripts x = witsScript x
```
Comment (nc6) [inline alonzo/impl/src/Cardano/Ledger/Alonzo/Tx.hs:644]: Duplicate?

File: alonzo/impl/src/Cardano/Ledger/Alonzo/TxBody.hs (hunk 8 above)
Comment (nc6) [reply alonzo/impl/src/Cardano/Ledger/Alonzo/TxBody.hs]: Even if it won't, git keeps history! So we can always delete commented code

File: semantics/small-steps-test/small-steps-test.cabal [hunk 23]
```diff
@@ -33,6 +33,7 @@ library
                      , microlens-th
                      , mtl
                      , nothunks
+                     , prettyprinter
```
Comment (nc6) [inline semantics/small-steps-test/small-steps-test.cabal:36]: How is this related to this PR?

**Review by nc6 — CHANGES_REQUESTED** [review]
A bunch of minor things

File: alonzo/impl/src/Cardano/Ledger/Alonzo/PParams.hs (hunk 1 above)
Comment (TimSheard) [reply alonzo/impl/src/Cardano/Ledger/Alonzo/PParams.hs]: done

File: alonzo/impl/src/Cardano/Ledger/Alonzo/PParams.hs (hunk 2 above)
Comment (TimSheard) [reply alonzo/impl/src/Cardano/Ledger/Alonzo/PParams.hs]: I changed it to HashTuple, since all instances are tuples.

File: alonzo/impl/src/Cardano/Ledger/Alonzo/Tx.hs (hunk 13 above)
Comment (JaredCorduan) [reply alonzo/impl/src/Cardano/Ledger/Alonzo/Tx.hs]: my guess is that this was done to match the actual subscript in the spec, which is IMHO reasonable.

File: alonzo/impl/src/Cardano/Ledger/Alonzo/Tx.hs (hunk 3 above)
Comment (TimSheard) [reply alonzo/impl/src/Cardano/Ledger/Alonzo/Tx.hs:50]: I have done an even better job, adding new figures.

File: alonzo/impl/src/Cardano/Ledger/Alonzo/Tx.hs (hunk 4 above)
Comment (TimSheard) [reply alonzo/impl/src/Cardano/Ledger/Alonzo/Tx.hs]: fixed that.

File: alonzo/impl/src/Cardano/Ledger/Alonzo/Tx.hs (hunk 5 above)
Comment (TimSheard) [reply alonzo/impl/src/Cardano/Ledger/Alonzo/Tx.hs]: done

File: alonzo/impl/src/Cardano/Ledger/Alonzo/Tx.hs (hunk 6 above)
Comment (TimSheard) [reply alonzo/impl/src/Cardano/Ledger/Alonzo/Tx.hs]: fixed the comment

File: alonzo/impl/src/Cardano/Ledger/Alonzo/Tx.hs (hunk 7 above)
Comment (TimSheard) [reply alonzo/impl/src/Cardano/Ledger/Alonzo/Tx.hs]: done

File: alonzo/impl/src/Cardano/Ledger/Alonzo/TxBody.hs (hunk 8 above)
Comment (TimSheard) [reply alonzo/impl/src/Cardano/Ledger/Alonzo/TxBody.hs]: done

File: alonzo/impl/src/Cardano/Ledger/Alonzo/PParams.hs (hunk 9 above)
Comment (TimSheard) [reply alonzo/impl/src/Cardano/Ledger/Alonzo/PParams.hs:93]: Yes, one should always separate the imports from the code. Our files sometimes have a hundred or more lines of imports.  This allows you to find where the code starts with a glance.

File: alonzo/impl/src/Cardano/Ledger/Alonzo/PParams.hs (hunk 10 above)
Comment (TimSheard) [reply alonzo/impl/src/Cardano/Ledger/Alonzo/PParams.hs:431]: OK I added this documentation to the module
This module exports implementations of many of the functions outlined in the Alonzo specification.
     The link to source of the specification
       https://github.com/input-output-hk/cardano-ledger-specs/tree/master/alonzo/formal-spec
     The most recent version of the document can be found here:
       https://hydra.iohk.io/job/Cardano/cardano-ledger-specs/specs.alonzo-ledger/latest/download-by-type/doc-pdf/alonzo-changes
     The functions can be found in Figures in that document, and sections of this code refer to those figures.

File: alonzo/impl/src/Cardano/Ledger/Alonzo/PParams.hs (hunk 11 above)
Comment (TimSheard) [reply alonzo/impl/src/Cardano/Ledger/Alonzo/PParams.hs]: I fixed this.

File: alonzo/impl/src/Cardano/Ledger/Alonzo/PParams.hs (hunk 12 above)
Comment (TimSheard) [reply alonzo/impl/src/Cardano/Ledger/Alonzo/PParams.hs]: I think we need some discussion about these things. The Alonzo spec is hashing al sorts of stuff. Ideally we would have some practical way to ensure that users can construct equivalent hashes. This is going to need some thought.

File: alonzo/impl/src/Cardano/Ledger/Alonzo/Tx.hs (hunk 13 above)
Comment (TimSheard) [reply alonzo/impl/src/Cardano/Ledger/Alonzo/Tx.hs]: This is an attempt to mimic the subscripts that are used in the specification.

File: alonzo/impl/src/Cardano/Ledger/Alonzo/Tx.hs (hunk 14 above)
Comment (TimSheard) [reply alonzo/impl/src/Cardano/Ledger/Alonzo/Tx.hs]: fixed this

File: alonzo/impl/src/Cardano/Ledger/Alonzo/Tx.hs (hunk 15 above)
Comment (TimSheard) [reply alonzo/impl/src/Cardano/Ledger/Alonzo/Tx.hs]: just mimicing the specification.

File: alonzo/impl/src/Cardano/Ledger/Alonzo/Tx.hs (hunk 16 above)
Comment (TimSheard) [reply alonzo/impl/src/Cardano/Ledger/Alonzo/Tx.hs:42]: done

File: alonzo/impl/src/Cardano/Ledger/Alonzo/Tx.hs (hunk 17 above)
Comment (TimSheard) [reply alonzo/impl/src/Cardano/Ledger/Alonzo/Tx.hs]: done

File: alonzo/impl/src/Cardano/Ledger/Alonzo/Tx.hs (hunk 19 above)
Comment (TimSheard) [reply alonzo/impl/src/Cardano/Ledger/Alonzo/Tx.hs:562]: Certainly possibility

File: alonzo/impl/src/Cardano/Ledger/Alonzo/Tx.hs (hunk 20 above)
Comment (TimSheard) [reply alonzo/impl/src/Cardano/Ledger/Alonzo/Tx.hs]: fixed

File: alonzo/impl/src/Cardano/Ledger/Alonzo/Tx.hs (hunk 21 above)
Comment (TimSheard) [reply alonzo/impl/src/Cardano/Ledger/Alonzo/Tx.hs:647]: mimicing the specification.

File: alonzo/impl/src/Cardano/Ledger/Alonzo/Tx.hs (hunk 22 above)
Comment (TimSheard) [reply alonzo/impl/src/Cardano/Ledger/Alonzo/Tx.hs:644]: Trying to keep the code consistent with the spec.  The other thing to do would be to change the names of the Haskell accessor functions.

File: semantics/small-steps-test/small-steps-test.cabal (hunk 23 above)
Comment (TimSheard) [reply semantics/small-steps-test/small-steps-test.cabal:36]: It seems the tests won't build without it. See Issue 2093, and the transcript of the build here
https://github.com/input-output-hk/cardano-ledger-specs/issues/2093

File: alonzo/impl/src/Cardano/Ledger/Alonzo/Tx.hs (hunk 18 above)
Comment (TimSheard) [reply alonzo/impl/src/Cardano/Ledger/Alonzo/Tx.hs:492]: This function has to wait until we have the Plutus API. Just a note to remind us to fix it when we have that.

File: alonzo/impl/src/Cardano/Ledger/Alonzo/PParams.hs (hunk 12 above)
Comment (nc6) [reply alonzo/impl/src/Cardano/Ledger/Alonzo/PParams.hs]: Happy to have a discussion about this and work out a sensible abstraction, but let's not introduce it in an unrelated module in an unrelated PR! Please go ahead and create a ticket that explains the problem here and then we can think about how to solve it.

File: alonzo/impl/src/Cardano/Ledger/Alonzo/Tx.hs (hunk 22 above)
Comment (nc6) [reply alonzo/impl/src/Cardano/Ledger/Alonzo/Tx.hs:644]: Gotcha. I think in this case it might be best to change the spec, to be honest!

File: alonzo/impl/src/Cardano/Ledger/Alonzo/PParams.hs (hunk 12 above)
Comment (TimSheard) [reply alonzo/impl/src/Cardano/Ledger/Alonzo/PParams.hs]: I have removed this. instead we use the strategy that anything hashed is a Memobytes.

File: alonzo/impl/src/Cardano/Ledger/Alonzo/Tx.hs (hunk 13 above)
Comment (TimSheard) [reply alonzo/impl/src/Cardano/Ledger/Alonzo/Tx.hs]: Renamed things consistent with other Era

File: alonzo/impl/src/Cardano/Ledger/Alonzo/Tx.hs (hunk 22 above)
Comment (TimSheard) [reply alonzo/impl/src/Cardano/Ledger/Alonzo/Tx.hs:644]: I will make some changes to the spec when Andre has fixed up his changes.

**Comment by TimSheard** [comment]
This latest push does a few things
1) It cleans up TxWitness so that the subpart that must be hashed is a public Memobytes. I also added a new pattern FlatWitness that lets users access it in the old style where there we 5 fields. So one can use either view.

2) I tried to make everything that needed to be hashed a Memobytes

3) I did some renaming of fields to be consistent and added HasField instances identedical to earlier errors.

4) There are still two places where hashing need to be further addressed. in PParams and the function hashScriptData.  I have stubbed out these parts. I suggest we get this PR merged and go back to address those issues.

**Review by nc6 — APPROVED** [review]
Looks great, thanks!
